scripts/west_commands/runners/sctool.py

```diff
@@ -59,22 +62,41 @@ def _flash(self):
         if not build_conf['CONFIG_SOC_RISCV_TELINK_W91']:
             print('only Telink W91 chip is supported!')
             exit()
+        # get flash size
+        flash_size = hex(build_conf['CONFIG_FLASH_SIZE'] * 1024)
+        # get flash base address
+        if self.address is not None: # if user defined use it
+            print(f'User defined flash base address: {self.address}')
+            flash_base = self.address
+        else:
+            # calculate flash base address from configuration
+            if build_conf.getboolean('CONFIG_MCUBOOT'): # mcuboot
+                print('MCUboot app!')
+                flash_base = hex(build_conf['CONFIG_FLASH_BASE_ADDRESS'])
+            elif build_conf.getboolean('CONFIG_BOOTLOADER_MCUBOOT'): # MCU boot app
+                print('MCUboot bootloader app!')
+                flash_base = hex(build_conf['CONFIG_FLASH_BASE_ADDRESS'] +
+                                 build_conf['CONFIG_TELINK_W91_SLOT0_PARTITION_ADDR'])
+            else: # standalone app
+                print('Standalone app!')
+                flash_base = hex(build_conf['CONFIG_FLASH_BASE_ADDRESS'])
+        # get N22 partition address
+        n22_partition_addr = hex(build_conf['CONFIG_TELINK_W91_N22_PARTITION_ADDR'] + int(flash_base, 16))
         # download RAM
-        pre_erase = subprocess.Popen(['./sctool', '-p', self.usb_port, '-da', 'da/da.ram.bin', '--before', 'hw_reset', '--after', 'no_reset', '-b', '115200', 'upload_da'], cwd=self.sctool_path)
+        pre_erase = subprocess.Popen(['./sctool', '-p', self.usb_port, '-da', 'da/da.ram.bin', '-b', self.baude_rate, '--before', 'hw_reset', '--after', 'no_reset', 'upload_da'], cwd=self.sctool_path)
         if pre_erase.wait():
             exit()
-        # get flash size
-        flash_size = hex(build_conf['CONFIG_FLASH_SIZE'] * 1024)
         print('Flashing D25 FW...')
-        # erase flash
+        # erase flash if needed
         if self.erase:
-            erase = subprocess.Popen(['./sctool', '-p', self.usb_port, '-da', 'da/da.ram.bin', '-b', '115200', '--manual', 'flash_erase', '0x80000000', flash_size], cwd=self.sctool_path)
+            print('Erasing flash...')
+            erase = subprocess.Popen(['./sctool', '-p', self.usb_port, '-da', 'da/da.ram.bin', '--manual', 'flash_erase', flash_base, flash_size], cwd=self.sctool_path)
```

Comment:
flash_size seems to be fixed
flash_size = hex(build_conf['CONFIG_FLASH_SIZE'] * 1024)
but flash_base app depending... So sometimes we are out of flash
I think erase should be applied to all chip (at least that's the easiest way to overcome the issue).

Reply:
https://github.com/telink-semi/zephyr/commit/c0f67c66f5957bfc964cf1a998610df304ba2e22